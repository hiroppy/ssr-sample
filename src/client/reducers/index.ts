import { combineReducers } from 'redux';
import {
  reducer as pagesReducer,
  initialState as pagesInitialState,
  State as PagesState
} from './pages';
import {
  reducer as sagaPageReducer,
  initialState as sagaPageInitialState,
  State as SagaPageState
} from './sagaPage';

export type State = {
  pages: PagesState;
  sagaPage: SagaPageState;
};

export const initialState = {
  pages: pagesInitialState,
  sagaPage: sagaPageInitialState
};

export const rootReducer = combineReducers({
  pages: pagesReducer,
  sagaPage: sagaPageReducer
});
