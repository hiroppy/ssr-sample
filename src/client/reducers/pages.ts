import { Actions } from '../actions/pages';

export type State = {
  isLoadingCompletion: boolean;
  error: Error | null;
  baseUrl: string;
  env: typeof process.env.NODE_ENV;
};

export const initialState: State = {
  isLoadingCompletion: false,
  error: null,
  baseUrl: '',
  env: 'development'
};

export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case 'SET_ENV':
      return { ...state, env: action.payload.env };
    case 'RESET_PAGES_STATUS':
      return initialState;
    case 'SET_BASE_URL':
      return { ...state, baseUrl: action.payload.baseUrl };
    case 'LOAD_TOP_PAGE':
    case 'LOAD_SAGA_PAGE':
    case 'LOAD_APOLLO_PAGE':
      return { ...state, isLoadingCompletion: false };
    case 'LOAD_TOP_PAGE_SUCCESS':
    case 'LOAD_SAGA_PAGE_SUCCESS':
    case 'LOAD_APOLLO_PAGE_SUCCESS':
      return { ...state, isLoadingCompletion: true };
    case 'LOAD_TOP_PAGE_FAILURE':
    case 'LOAD_SAGA_PAGE_FAILURE':
    case 'LOAD_APOLLO_PAGE_FAILURE':
      return { ...state, isLoadingCompletion: true, error: action.payload.err };
    default:
      return state;
  }
};
