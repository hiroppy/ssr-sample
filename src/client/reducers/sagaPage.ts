import { Actions } from '../actions/fetchSaga';
import { Samples } from '../../server/responseSchema';

export type State = {
  samples: Samples;
};

export const initialState: State = {
  samples: []
};

export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case 'FETCH_SAGA_CODE_SUCCESS':
      return { ...state, samples: action.payload.samples };
    case 'FETCH_SAGA_CODE':
    case 'FETCH_SAGA_CODE_FAILURE':
    default:
      return state;
  }
};
