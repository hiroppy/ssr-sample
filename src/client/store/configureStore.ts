import { createStore, applyMiddleware, Store } from 'redux';
import createSagaMiddleware, { SagaMiddleware } from 'redux-saga';
import { createBrowserHistory, createMemoryHistory } from 'history';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { rootReducer } from '../reducers';
import { rootSaga } from '../sagas';

export const history = process.env.IS_BROWSER ? createBrowserHistory() : createMemoryHistory();

const createEnhancer = (sagaMiddleware: ReturnType<typeof createSagaMiddleware>) => {
  const composeEnhancers = composeWithDevTools({});

  return composeEnhancers(applyMiddleware(sagaMiddleware));
};

export const configureStore = (
  preloadedState: Record<string, any> = {}
): {
  store: Store;
  runSaga: () => Promise<SagaMiddleware<typeof rootSaga>['run']>;
} => {
  const sagaMiddleware = createSagaMiddleware();
  const enhancer = createEnhancer(sagaMiddleware);
  const store = createStore(rootReducer, preloadedState, enhancer);
  const runSaga = async () => {
    return sagaMiddleware.run(rootSaga).toPromise();
  };

  // for client-side
  if (process.env.IS_BROWSER) {
    runSaga();
  }

  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('../reducers', async () => {
      const { rootReducer } = await import(/* webpackMode: "eager" */ '../reducers');

      store.replaceReducer(rootReducer);
    });
  }

  return { store, runSaga };
};
