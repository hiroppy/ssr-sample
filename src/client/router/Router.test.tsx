import * as React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import configureStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import { rootReducer } from '../reducers';
import { Router } from '.';

let state: any;
let store: any;

jest.mock('./routes.ts', () => {
  return {
    LoadableTopPage: () => <p>top</p>,
    LoadableSagaPage: () => <p>saga</p>,
    LoadableApolloPage: () => <p>apollo</p>,
    LoadableNotFoundPage: () => <p>404</p>
  };
});

beforeEach(() => {
  state = createStore(rootReducer).getState();
  store = configureStore()(state);
});

test('should render Top page', () => {
  const tree = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <Router />
      </MemoryRouter>
    </Provider>
  );

  expect(tree.asFragment()).toMatchSnapshot();
});

test('should render Saga page', () => {
  const tree = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={[{ pathname: '/saga', key: 'testKey' }]}>
        <Router />
      </MemoryRouter>
    </Provider>
  );

  expect(tree.asFragment()).toMatchSnapshot();
});

test('should render Apollo page', () => {
  const tree = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={[{ pathname: '/apollo', key: 'testKey' }]}>
        <Router />
      </MemoryRouter>
    </Provider>
  );

  expect(tree.asFragment()).toMatchSnapshot();
});

test('should render 404 page', () => {
  const tree = render(
    <Provider store={store}>
      <MemoryRouter initialEntries={[{ pathname: '/noPath', key: 'testKey' }]}>
        <Router />
      </MemoryRouter>
    </Provider>
  );

  expect(tree.asFragment()).toMatchSnapshot();
});
