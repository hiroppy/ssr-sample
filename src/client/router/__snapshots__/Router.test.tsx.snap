// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 404 page 1`] = `
<DocumentFragment>
  <header
    class="sc-AykKD buFqre"
  >
    <a
      class="sc-AykKC fSVTkQ"
      href="/"
    >
      &lt;
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/saga"
    >
      saga
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/apollo"
    >
      apollo
    </a>
    <span>
      development
    </span>
  </header>
  <div
    class="sc-AykKE ffQxyb"
  >
    <p>
      404
    </p>
  </div>
</DocumentFragment>
`;

exports[`should render Apollo page 1`] = `
<DocumentFragment>
  <header
    class="sc-AykKD buFqre"
  >
    <a
      class="sc-AykKC fSVTkQ"
      href="/"
    >
      &lt;
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/saga"
    >
      saga
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/apollo"
    >
      apollo
    </a>
    <span>
      development
    </span>
  </header>
  <div
    class="sc-AykKE ffQxyb"
  >
    <p>
      apollo
    </p>
  </div>
</DocumentFragment>
`;

exports[`should render Saga page 1`] = `
<DocumentFragment>
  <header
    class="sc-AykKD buFqre"
  >
    <a
      class="sc-AykKC fSVTkQ"
      href="/"
    >
      &lt;
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/saga"
    >
      saga
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/apollo"
    >
      apollo
    </a>
    <span>
      development
    </span>
  </header>
  <div
    class="sc-AykKE ffQxyb"
  >
    <p>
      saga
    </p>
  </div>
</DocumentFragment>
`;

exports[`should render Top page 1`] = `
<DocumentFragment>
  <header
    class="sc-AykKD buFqre"
  >
    <a
      class="sc-AykKC fSVTkQ"
      href="/"
    >
      &lt;
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/saga"
    >
      saga
    </a>
    <a
      class="sc-AykKC fSVTkQ"
      href="/apollo"
    >
      apollo
    </a>
    <span>
      development
    </span>
  </header>
  <div
    class="sc-AykKE ffQxyb"
  >
    <p>
      top
    </p>
  </div>
</DocumentFragment>
`;
